connectors/sources/graphql.py

```diff
@@ -0,0 +1,449 @@
+        if node.arguments:
+            for arg in node.arguments:
+                self.fields_dict[node.name.value].append(arg.name.value)
+                if isinstance(arg.value, VariableNode):
+                    self.variables_dict[node.name.value].append(arg.value.name.value)
+
+
+class UnauthorizedException(Exception):
+    pass
+
+
+class GraphQLClient:
+    def __init__(self, configuration):
+        self._sleeps = CancellableSleeps()
+        self.configuration = configuration
+        self._logger = logger
+        self.graphql_query = self.configuration["graphql_query"]
+        self.url = self.configuration["http_endpoint"]
+        self.graphql_object_list = self.configuration["graphql_object_list"]
+        self.http_method = self.configuration["http_method"]
+        self.authentication_method = self.configuration["authentication_method"]
+        self.pagination_enabled = self.configuration["pagination_enabled"]
+        self.variables = {}
+        self.headers = {}
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    @cached_property
+    def session(self):
+        timeout = aiohttp.ClientTimeout(total=self.configuration["connection_timeout"])
+        if self.authentication_method == "bearer":
+            self.headers.update(
+                {
+                    "Authorization": f"Bearer {self.configuration['token']}",
+                }
+            )
+            return aiohttp.ClientSession(
+                headers=self.headers,
+                timeout=timeout,
+                raise_for_status=True,
+            )
+        elif self.authentication_method == "basic":
+            basic_auth = aiohttp.BasicAuth(
+                login=self.configuration["username"],
+                password=self.configuration["password"],
+            )
+            return aiohttp.ClientSession(
+                auth=basic_auth,
+                headers=self.headers,
+                timeout=timeout,
+                raise_for_status=True,
+            )
+        else:
+            return aiohttp.ClientSession(
+                headers=self.headers,
+                timeout=timeout,
+                raise_for_status=True,
+            )
+
+    def extract_graphql_data_items(self, key, data):
+        """Returns sub objects from the response based on graphql_object_list
+
+        Args:
+            key (string): Key of data.
+            data (dict): data to extract
+
+        Yields:
+            dictionary/list: Documents from the response
+        """
+        if key in self.graphql_object_list:
+            yield data
+        if isinstance(data, dict):
+            for key, item in data.items():
+                yield from self.extract_graphql_data_items(key=key, data=item)
+        if isinstance(data, list):
+            for doc in data:
+                yield from self.extract_graphql_data_items(key=None, data=doc)
+
+    async def paginated_call(self, graphql_query):
```

Comment:
This function needs more debug logging to troubleshoot configuration issues. We need to be able to log what pagination type is being used, what are the values of pagination variables